pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -31,10 +41,82 @@ import (
 	"istio.io/istio/pkg/log"
 )
 
-// Endpoints implements MeshDiscovery.Endpoints()
-func Endpoints(ds *v1.DiscoveryService, serviceClusters []string) *xdsapi.DiscoveryResponse {
+// EDS returns the list of endpoints (IP:port and in future labels) associated with a real
+// or virtual service. A virtual service is a subset of a real service, selected using labels.
+//
+// The source of info is a list of service registries.
+//
+// Primary event is an endpoint creation/deletion. Once the event is fired, EDS needs to
+// find the list of services associated with the endpoint.
+//
+// In case of k8s, Endpoints event is fired when the endpoints are added to service - typically
+// after readiness check. At that point we have the 'real' Service. The Endpoint includes a list
+// of port numbers and names.
+//
+// For the virtual case, the Pod referenced in the Endpoint must be looked up, and pod checked
+// for labels.
+//
+// In addition, ExternalEndpoint includes IPs and labels directly and can be directly processed.
+//
+// TODO: for selector-less services (mesh expansion), skip pod processing
+// TODO: optimize the code path for ExternalEndpoint, no additional processing needed
+// TODO: if a service doesn't have split traffic - we can also skip pod and lable processing
+// TODO: efficient label processing. In alpha3, the destination policies are set per service, so
+// we may only need to search in a small list.
+
+var (
+	// It doesn't appear the logging framework has support for per-component logging, yet
+	// we need to be able to debug EDS2.
+	edsDebug = len(os.Getenv("PILOT_DEBUG_EDS")) == 0
+
+	edsClusterMutex sync.Mutex
+	edsClusters     = map[string]*EdsCluster{}
+	version         = 1
+)
+
+// EdsCluster tracks eds-related info for monitored clusters. In practice it'll include
+// all clusters until we support on-demand cluster loading.
+type EdsCluster struct {
+	// mutex protects changes to this cluster
+	mutex sync.Mutex
+
+	// EdsClients keeps track of all nodes monitoring the cluster.
+	EdsClients map[string]*EdsConnection
+
+	LoadAssignments *xdsapi.ClusterLoadAssignment
+
+	// FirstUse is the time the cluster was first used, for debugging
+	FirstUse time.Time
+
+	// NonEmptyTime is the time the cluster first had a non-empty set of endpoints
+	NonEmptyTime time.Time
+
+	// The discovery service this cluster is associated with.
+	discovery *DiscoveryServer
+}
+
+// TODO: add prom metrics !
+
+// EdsConnection represents a streaming grpc connection from an envoy server.
+// This is primarily intended for supporting push, but also for debug and statusz.
+type EdsConnection struct {
+	PeerAddr string
+
+	// current list of clusters monitored by the client
+	Clusters []string
```

Comment:
Why not just have an array of EDSClusters ?

Reply:
It's the list of cluster names that are monitored in that connection. For debug it would look to verbose. For other uses - so far it doesn't seem to be needed. If we do find a need - we can add or change.